cmake/create_test.cmake

```diff
@@ -31,26 +24,75 @@ function(ginkgo_set_test_target_properties test_target_name)
     if (GINKGO_COMPILING_DPCPP_TEST AND GINKGO_DPCPP_SINGLE_MODE)
         target_compile_definitions(${test_target_name} PRIVATE GINKGO_DPCPP_SINGLE_MODE=1)
     endif()
-    if (GINKGO_CHECK_CIRCULAR_DEPS)
+    if(GINKGO_CHECK_CIRCULAR_DEPS)
         target_link_libraries(${test_target_name} PRIVATE "${GINKGO_CIRCULAR_DEPS_FLAGS}")
     endif()
-    if (set_properties_MPI_SIZE)
-        if(NOT TARGET gtest_mpi_main)
-            ginkgo_create_gtest_mpi_main()
-        endif()
-        set(gtest_main gtest_mpi_main MPI::MPI_CXX)
+    if(set_properties_MPI_SIZE)
+        target_link_libraries(${test_target_name} PRIVATE ginkgo_gtest_main_mpi${test_library_suffix})
     else()
-        set(gtest_main GTest::Main)
+        target_link_libraries(${test_target_name} PRIVATE ginkgo_gtest_main${test_library_suffix})
     endif()
     target_compile_features(${test_target_name} PUBLIC cxx_std_14)
     target_compile_options(${test_target_name} PRIVATE $<$<COMPILE_LANGUAGE:CXX>:${GINKGO_COMPILER_FLAGS}>)
     target_include_directories(${test_target_name} PRIVATE ${Ginkgo_BINARY_DIR} ${set_properties_ADDITIONAL_INCLUDES})
-    target_link_libraries(${test_target_name} PRIVATE ginkgo ${gtest_main} GTest::GTest ${set_properties_ADDITIONAL_LIBRARIES})
+    target_link_libraries(${test_target_name} PRIVATE ginkgo GTest::GTest ${set_properties_ADDITIONAL_LIBRARIES})
+endfunction()
+
+function(ginkgo_add_resource_requirement test_name)
+    cmake_parse_arguments(PARSE_ARGV 1 add_rr "${gko_test_option_args}" "${gko_test_single_args}" "")
+    if(add_rr_NO_RESOURCES)
+        return()
+    endif()
+
+    if (NOT add_rr_RESOURCE_TYPE)
+        message(FATAL_ERROR "Need to provide resource type used by test.")
+    endif ()
+
+    if(add_rr_RESOURCE_TYPE STREQUAL "ref")
+        set(single_resource "cpu:1")
+    elseif(add_rr_RESOURCE_TYPE STREQUAL "cpu")
+        if(NOT add_rr_RESOURCE_LOCAL_CORES)
+            set(add_rr_RESOURCE_LOCAL_CORES ${GINKGO_TEST_OMP_PARALLELISM})
+        endif()
+        if(NOT add_rr_RESOURCE_LOCAL_CORES MATCHES "^[0-9]+")
+            message(FATAL_ERROR "Resource specification is invalid: RESOURCE_LOCAL_CORES=${add_rr_RESOURCE_LOCAL_CORES}")
+        endif()
+
+        set(single_resource "cpu:${add_rr_RESOURCE_LOCAL_CORES}")
+    elseif(add_rr_RESOURCE_TYPE MATCHES "^(cudagpu|hipgpu|sycl)$")
+        if(NOT add_rr_RESOURCE_PERCENTAGE)
+            set(add_rr_RESOURCE_PERCENTAGE 25)
+        endif()
+        if(add_rr_MPI_SIZE GREATER 1)
+            set(add_rr_RESOURCE_PERCENTAGE 100)
+        endif()
```

Comment:
it's unnecessary.

Reply:
I think we had some issues with MPI tests being extremely slow if oversubscribing a single GPU. Since we have multi-GPU machines available for each architecture where we test MPI, we can still keep this restriction.

Reply:
let's keep this discussion in a single place: https://github.com/ginkgo-project/ginkgo/pull/1373#discussion_r1315754500